### DISCUSSION TITLE
Support for Everlight 12-23C SmartLED

### DISCUSSION THREAD

**Discussion by jcanderle** [Makuna/NeoPixelBus#507] [Q&A] [answered]
Hi,

I have SmartLED  [Everlight 12-23C](https://en.everlight.com/wp-content/plugins/ItemRelationship/product_files/pdf/12-23C-RSGHBHW-5V01-2C.pdf) which has slightly different timings from WS2812 LEDs.
I am looking to add or change some helper methods in /NeoPixelBus/src/internal/NeoEsp32I2sMethod.h
How can I change the timings to fit the following table:

Item | Description | Typical  | Allowance 
------------ | ------------- | ------------- | -------------
T0H | 0 code‧High-level time | 300ns | ±80ns 
T0L | 0 code‧Low-level time | 900ns | ±80ns 
T1H | 1 code‧High-level time | 900ns | ±80ns 
T1L | 1 code‧Low-level time | 300ns | ±80ns 
RES | Reset time | >50us |

**Comment by Makuna** [comment, marked as answer]
I2s methods do not have the means to set the pulse timing specifically.  They are sent using a ratio of sorts, either 1/4 of cycle width "on" or 3/4 of cycle width is "on".  The timing you listed above already meets this criteria (300ns is 1/4 of 1200ns, 900ns is 3/4 of 1200ns).  Further, all the methods already have a reset time of > 50us.

Further, the i2s doesn't produce the most consistent pulse widths either, so +-50ns is very common.

The actual cycle width is set to 1250us (the most common spec) which still should produce a signal that meets the spec you listed above.  Did you try the already available methods?

The RMT methods do provide specific timing and the closest one to the above would be the NeoEsp32Rmt1Ws2811Method.

**Reply by Makuna** [reply]
BTW, if need to stick with i2s rather than use RMT, then you can try the following to reduce the total cycle width to 1200ns and let me know if it helps.

Modify the I2sSampleRate value for the  NeoEsp32I2sSpeedWs2812x class in NeoEsp32I2sMethod.h from 100000 to 104000 like ... 
```
class NeoEsp32I2sSpeedWs2812x
{
public:
    const static uint32_t I2sSampleRate =  104000; // original was 100000;
    const static uint16_t ByteSendTimeUs = 10;
    const static uint16_t ResetTimeUs = 300;
};
```
